src/dsp_tools/commands/xmlupload/transform_resource_and_values_for_upload.py

```diff
@@ -242,167 +225,257 @@ def _make_bitstream_file_value(bitstream_info: BitstreamInfo) -> dict[str, Any]:
 
 
-def _make_boolean_prop(
-    prop: XMLProperty, res_bn: BNode, prop_name: URIRef, permissions_lookup: dict[str, Permissions]
+def _make_simple_prop_graph(
+    prop: XMLProperty,
+    res_bn: BNode,
+    prop_name: URIRef,
+    prop_type_info: RDFPropTypeInfo,
+    transformer: Callable[[InputTypes], Literal],
+    permissions_lookup: dict[str, Permissions],
 ) -> Graph:
     g = Graph()
-    for value in prop.values:
-        boolean_value = _make_boolean_value(value, prop_name, res_bn, permissions_lookup)
-        g += boolean_value.as_graph()
+    for val in prop.values:
+        transformed_val = transformer(val.value)
+        resolved_permission = _resolve_permission(val.permissions, permissions_lookup)
+        transformed = TransformedValue(
+            value=transformed_val, prop_name=prop_name, permissions=resolved_permission, comment=val.comment
+        )
+        g += _make_simple_value_graph(transformed, res_bn, prop_type_info)
     return g
 
 
-def _make_boolean_value(
-    value: XMLValue, prop_name: URIRef, res_bn: BNode, permissions_lookup: dict[str, Permissions]
-) -> BooleanValueRDF:
-    s = assert_is_string(value.value)
-    as_bool = transform_boolean(s)
-    permission_literal = None
-    if permission_str := _get_permission_str(value.permissions, permissions_lookup):
-        permission_literal = Literal(permission_str)
-    return BooleanValueRDF(
-        resource_bn=res_bn,
-        prop_name=prop_name,
-        value=Literal(as_bool),
-        permissions=permission_literal,
-        comment=Literal(value.comment) if value.comment else None,
-    )
+def _make_simple_value_graph(
+    val: TransformedValue,
+    res_bn: BNode,
+    prop_type_info: RDFPropTypeInfo,
+) -> Graph:
+    val_bn = BNode()
+    g = _add_optional_triples(val_bn, val.permissions, val.comment)
+    g.add((res_bn, val.prop_name, val_bn))
+    g.add((val_bn, RDF.type, prop_type_info.knora_type))
+    g.add((val_bn, prop_type_info.knora_prop, val.value))
+    return g
 
 
-def _make_integer_prop(
-    prop: XMLProperty, res_bn: BNode, prop_name: URIRef, permissions_lookup: dict[str, Permissions]
+def _make_list_prop_graph(
+    prop: XMLProperty,
+    res_bn: BNode,
+    prop_name: URIRef,
+    permissions_lookup: dict[str, Permissions],
+    listnode_lookup: dict[str, str],
 ) -> Graph:
     g = Graph()
-    for value in prop.values:
-        int_value = _make_integer_value(value, prop_name, res_bn, permissions_lookup)
-        g += int_value.as_graph()
+    for val in prop.values:
+        s = assert_is_string(val.value)
+        if not (iri_str := listnode_lookup.get(s)):
+            msg = (
+                f"Could not resolve list node ID '{s}' to IRI. "
+                f"This is probably because the list node '{s}' does not exist on the server."
+            )
+            raise BaseError(msg)
+
+        resolved_permission = _resolve_permission(val.permissions, permissions_lookup)
```

Comment:
After the list node is resolved, the value can be serialised with the generic function.